oak-run/src/main/java/org/apache/jackrabbit/oak/indexversion/IndexVersionOperation.java

```diff
@@ -0,0 +1,197 @@
+            indexVersionOperationList = Collections.emptyList();
+        }
+        return indexVersionOperationList;
+    }
+
+    /*
+        returns IndexNameObjects in descending order of version i.e from latest to oldest
+    */
+    private static List<IndexName> getSortedVersionedIndexNameObjects(List<IndexName> indexNameObjectList) {
+        indexNameObjectList.sort(new Comparator<IndexName>() {
+            @Override
+            public int compare(IndexName indexNameObj1, IndexName indexNameObj2) {
+                return indexNameObj2.compareTo(indexNameObj1);
+            }
+        });
+        return indexNameObjectList;
+    }
+
+    /**
+     * @param indexVersionOperations
+     * @return true if the IndexVersionOperation list passes following criteria.
+     * For merging indexes we need baseIndex and latest custom index.
+     * So we first validate that if there are custom indexes than OOTB index with same product must be marked with DELETE_HIDDEN_AND_DISABLE
+     */
+    private static boolean isValidIndexVersionOperationList(List<IndexVersionOperation> indexVersionOperations) {
+        boolean isValid = false;
+        IndexVersionOperation lastNoopOperationIndexVersion = null;
+        IndexVersionOperation indexWithDeleteHiddenOp = null;
+        for (IndexVersionOperation indexVersionOperation : indexVersionOperations) {
+            if (indexVersionOperation.getOperation() == Operation.NOOP) {
+                lastNoopOperationIndexVersion = indexVersionOperation;
+            }
+            if (indexVersionOperation.getOperation() == Operation.DELETE_HIDDEN_AND_DISABLE) {
+                indexWithDeleteHiddenOp = indexVersionOperation;
+            }
+        }
+        if (lastNoopOperationIndexVersion.getIndexName().getCustomerVersion() == 0) {
+            isValid = true;
+        } else if (lastNoopOperationIndexVersion.getIndexName().getCustomerVersion() != 0) {
+            if (indexWithDeleteHiddenOp != null
+                    && lastNoopOperationIndexVersion.getIndexName().getProductVersion() == indexWithDeleteHiddenOp.getIndexName().getProductVersion()) {
+                isValid = true;
+            }
+        }
+        if (!isValid) {
+            LOG.info("IndexVersionOperation List is not valid for index {}", lastNoopOperationIndexVersion.getIndexName().getNodeName());
+        }
+        return isValid;
+    }
+
+    private static void removeDisabledCustomIndexes(NodeState indexDefParentNode,
+                                                    List<IndexName> indexNameObjectList) {
+        for (int i = 0; i < indexNameObjectList.size(); i++) {
+            NodeState indexNode = indexDefParentNode
+                    .getChildNode(PathUtils.getName(indexNameObjectList.get(i).getNodeName()));
+            if (indexNode.getProperty("type") != null && indexNode.getProperty("type").getValue(Type.STRING).equals("disabled")) {
```

Comment:
It is fine like it is, I would keep it. But just a general remark about methods that can return null: you can often simplify the logic by reversing the comparison. I'm not sure if it would work here:

    if ("disabled".equals(indexNode.getProperty("type").getValue(Type.STRING)) {